labgrid/autoinstall/main.py

```diff
@@ -0,0 +1,203 @@
+"""The autoinstall.main module runs an installation script automatically on multiple targets."""
+import ast
+import argparse
+import logging
+import sys
+import traceback
+import multiprocessing
+import textwrap
+from contextlib import closing
+from socket import gethostname
+from time import sleep
+
+import attr
+
+from .. import Environment, target_factory
+from ..stepreporter import StepReporter
+from ..exceptions import NoResourceFoundError
+
+
+class Handler(multiprocessing.Process):
+    def __init__(self, env, name):
+        super().__init__(name=name)
+        self.env = env
+        self.config = env.config
+        self.name = name
+
+        self.context = {
+            'name': self.name,
+            'env': self.env,
+            'config': self.config,
+        }
+        self.context.update(target_factory.resources)
+        self.context.update(target_factory.drivers)
+
+    def _get_function(self, name, context):
+        snippet = self.config.data['autoinstall'].get(name)
+        if not snippet:
+            return None
+
+        code = 'def {}():\n{}'.format(name, textwrap.indent(snippet, ' '))
+        tree = ast.parse(code, filename=self.env.config_file)
+        ast.increment_lineno(tree, snippet.start_mark.line)
+        co = compile(tree, filename=self.env.config_file, mode='exec')
+
+        stage = {}
+        exec(co, context, stage)
+        return stage[name]
+
+    def _get_setup_function(self):
+        context = self.context
+        context['log'] = self.log.getChild('setup')
+        setup = self._get_function('setup', self.context)
+        if setup is None:
+            def setup():
+                pass
+        return setup
+
+    def _get_handler_function(self):
+        context = self.context.copy()
+        context['log'] = self.log.getChild('handler')
+        handler = self._get_function('handler', context)
+        return handler
+
+    def _get_initial_resource(self):
+        cls = self.config.data['autoinstall'].get('initial-resource')
+        if not cls:
+            return None
+
+        return self.target.get_resource(self.context[cls], await=False)
+
+    def run(self):
+        self.log = logging.getLogger(self.name)
+        self.log.info("creating new handler (PID {})".format(self.pid))
```

Comment:
[Logging Optimizations](https://docs.python.org/3/howto/logging.html#optimization) mentions that values should be passed via lazy interpolation, to avoid the formatting cost if the message is not emitted, e.g. 
```
self.log.info("creating new handler (PID %s)", self.pid)
```
instead of
```
self.log.info("creating new handler (PID {})".format(self.pid))
```